### PR TITLE
Feat: add webhook token in application

### PR DESCRIPTION
**Pull request by FogDong**
Signed-off-by: FogDong <dongtianxin.tx@alibaba-inc.com>

### Description of your changes

add webhook token in application

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

/cc @barnettZQG @wonderflow

### CODE REVIEW

File: pkg/apiserver/rest/webservice/application.go [hunk 1]
```diff
@@ -103,6 +103,19 @@ func (c *applicationWebService) GetWebService() *restful.WebService {
 		Returns(200, "", apis.ApplicationBase{}).
 		Returns(400, "", bcode.Bcode{}).
 		Writes(apis.ApplicationBase{}))
+
+	ws.Route(ws.GET("/{name}/webhook/{token}").To(c.handleApplicationWebhook).
```
Comment (wonderflow) [inline pkg/apiserver/rest/webservice/application.go:107]:
```suggestion
	ws.Route(ws.POST("/webhook/{token}").To(c.handleApplicationWebhook).
```

File: pkg/apiserver/model/application.go [hunk 2]
```diff
@@ -24,19 +24,20 @@ import (
...
-	Icon        string            `json:"icon"`
-	Labels      map[string]string `json:"labels,omitempty"`
+	Name         string            `json:"name"`
+	Alias        string            `json:"alias"`
+	Project      string            `json:"project"`
+	Namespace    string            `json:"namespace"`
+	Description  string            `json:"description"`
+	Icon         string            `json:"icon"`
+	Labels       map[string]string `json:"labels,omitempty"`
+	WebhookToken string            `json:"webhookToken"`
```
Comment (wonderflow) [inline pkg/apiserver/model/application.go:40]: do we still need a field here?

Comment (FogDong) [reply pkg/apiserver/model/application.go:40]: Or we need to get the token for every application in a LIST request.

**Review by wonderflow — APPROVED** [review]
good job

File: pkg/apiserver/model/application.go (hunk 2 above)
Comment (barnettZQG) [reply pkg/apiserver/model/application.go:40]: I think it can be deleted.

File: pkg/apiserver/model/application.go [hunk 3]
```diff
@@ -253,3 +266,29 @@ func (a *ApplicationRevision) Index() map[string]string {
 	}
 	return index
 }
+
+// ApplicationWebhook is the model for webhook request
+type ApplicationWebhook struct {
+	Model
+	AppPrimaryKey string `json:"appPrimaryKey"`
+	Token         string `json:"token"`
```
Comment (barnettZQG) [inline pkg/apiserver/model/application.go:274]: Including workflow name? The user doesn't care about the app name, so doesn't care about the workflow name.

File: pkg/apiserver/model/model.go [hunk 4]
```diff
@@ -98,6 +98,17 @@ func (j *JSONStruct) JSON() string {
 	return string(b)
 }
 
+// MergeFrom merges json struct from another struct
+func (j *JSONStruct) MergeFrom(from JSONStruct) *JSONStruct {
+	if j == nil {
+		return &from
+	}
+	for k, v := range from {
```
Comment (barnettZQG) [inline pkg/apiserver/model/model.go:106]: This only supports one dimension, the best recursive patch?

Comment (barnettZQG) [reply pkg/apiserver/model/model.go:106]: I remember having implemented the tool methods available.

File: pkg/apiserver/rest/apis/v1/types.go [hunk 5]
```diff
@@ -292,14 +291,15 @@ func (e EnvBindingList) ContainTarget(name string) bool {
...
-	Labels      map[string]string `json:"labels,omitempty"`
+	Name         string            `json:"name"`
+	Alias        string            `json:"alias"`
+	Project      *ProjectBase      `json:"project"`
+	Description  string            `json:"description"`
+	CreateTime   time.Time         `json:"createTime"`
+	UpdateTime   time.Time         `json:"updateTime"`
+	Icon         string            `json:"icon"`
+	Labels       map[string]string `json:"labels,omitempty"`
+	WebhookToken string            `json:"webhookToken"`
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:302]: have only one Token?

Comment (barnettZQG) [reply pkg/apiserver/rest/apis/v1/types.go:302]: It is recommended to provide an independent webhook query API,  we should implement deletion and creation API in the follow-up.

Comment (FogDong) [reply pkg/apiserver/rest/apis/v1/types.go:302]: A application should have multiple webhooks?
The webhook is an independent API

File: pkg/apiserver/model/application.go (hunk 3 above)
Comment (FogDong) [reply pkg/apiserver/model/application.go:274]: It‘s a optional param, default to default workfkow. We need to choose workflow when we’re depoying.

File: pkg/apiserver/model/application.go [hunk 6]
```diff
@@ -249,3 +261,34 @@ func (a *ApplicationRevision) Index() map[string]string {
...
+// ApplicationTrigger is the model for trigger
+type ApplicationTrigger struct {
+	BaseModel
+	AppPrimaryKey string `json:"appPrimaryKey"`
+	WorkflowName  string `json:"workflowName"`
+	Name          string `json:"name"`
+	Alias         string `json:"alias,omitempty"`
+	Description   string `json:"description,omitempty"`
+	Token         string `json:"token"`
+	Type          string `json:"type"`
```
Comment (barnettZQG) [inline pkg/apiserver/model/application.go:274]: 应该少了一个参数，这个Type取值 webhook 时，他还需要指定 Payload Type。可以设置一个扩展参数？

File: pkg/apiserver/model/application.go [hunk 7]
```diff
@@ -249,3 +261,34 @@ func (a *ApplicationRevision) Index() map[string]string {
...
+func (w *ApplicationTrigger) PrimaryKey() string {
+	return w.Token
+}
+
+// Index return custom index
+func (w *ApplicationTrigger) Index() map[string]string {
+	index := make(map[string]string)
+	if w.Token != "" {
+		index["token"] = w.Token
+	}
```
Comment (barnettZQG) [inline pkg/apiserver/model/application.go:292]: name，type 都应该设置为索引

File: pkg/apiserver/rest/apis/v1/types.go [hunk 8]
```diff
@@ -386,11 +410,12 @@ type ComponentSelector struct {
...
-	EnvBindings     []string                `json:"envBindings"`
-	Status          string                  `json:"status"`
-	ApplicationType string                  `json:"applicationType"`
-	ResourceInfo    ApplicationResourceInfo `json:"resourceInfo"`
+	Policies        []string                  `json:"policies"`
+	EnvBindings     []string                  `json:"envBindings"`
+	Status          string                    `json:"status"`
+	ApplicationType string                    `json:"applicationType"`
+	ResourceInfo    ApplicationResourceInfo   `json:"resourceInfo"`
+	Triggers        []*ApplicationTriggerBase `json:"triggers"`
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:418]: 独立一个API吧。先实现list查询即可。

File: pkg/apiserver/rest/webservice/webhook.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+		Reads(apis.HandleApplicationWebhookRequest{}).
+		Returns(200, "", apis.ApplicationDeployResponse{}).
+		Returns(400, "", bcode.Bcode{}).
+		Writes(apis.ApplicationDeployResponse{}))
+	return ws
+}
+
+func (c *webhookWebService) handleApplicationWebhook(req *restful.Request, res *restful.Response) {
+	var webhookReq apis.HandleApplicationWebhookRequest
+	if err := req.ReadEntity(&webhookReq); err != nil {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/webservice/webhook.go:63]: 这里应该根据 Payload Type 来解析。

File: pkg/apiserver/rest/apis/v1/types.go [hunk 10]
```diff
@@ -332,14 +331,41 @@ type CreateApplicationRequest struct {
...
+
+// ApplicationTriggerBase application trigger base model
+type ApplicationTriggerBase struct {
+	Name         string `json:"name"`
+	Alias        string `json:"alias"`
+	Description  string `json:"description"`
+	WorkflowName string `json:"workflowName"`
+	Type         string `json:"type"`
+	PayloadType  string `json:"payloadType"`
+	Token        string `json:"token"`
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:360]: createTime/updateTime

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**